presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -0,0 +1,269 @@
+    private final HttpClient httpClient;
+    private final TableWriteInfo tableWriteInfo;
+    private final RemoteTaskStats stats;
+    private final List<TaskSource> sources;
+    private TaskInfo taskInfo;
+    private final Executor executor;
+    private final ScheduledExecutorService updateScheduledExecutor;
+    @GuardedBy("this")
+    private ScheduledFuture<?> scheduledFuture;
+    private final JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec;
+    private final JsonCodec<PlanFragment> planFragmentCodec;
+    private final JsonCodec<TaskInfo> taskInfoCodec;
+
+    public NativeExecutionTask(
+            Session session,
+            TaskId taskId,
+            String nodeId,
+            URI location,
+            PlanFragment planFragment,
+            List<TaskSource> sources,
+            OutputBuffers outputBuffers,
+            HttpClient httpClient,
+            TableWriteInfo tableWriteInfo,
+            RemoteTaskStats stats,
+            Executor executor,
+            ScheduledExecutorService updateScheduledExecutor,
+            JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec,
+            JsonCodec<PlanFragment> planFragmentCodec,
+            JsonCodec<TaskInfo> taskInfoCodec)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(taskId, "taskId is null");
+        requireNonNull(nodeId, "nodeId is null");
+        requireNonNull(location, "location is null");
+        requireNonNull(planFragment, "planFragment is null");
+        requireNonNull(sources, "sources is null");
+        requireNonNull(outputBuffers, "outputBuffers is null");
+        requireNonNull(httpClient, "httpClient is null");
+
+        try (SetThreadName ignored = new SetThreadName("HttpRemoteTask-%s", taskId)) {
+            this.taskId = taskId;
+            this.session = session;
+            this.planFragment = planFragment;
+            this.outputBuffers.set(outputBuffers);
+            this.httpClient = httpClient;
+            this.tableWriteInfo = tableWriteInfo;
+            this.stats = stats;
+            this.sources = sources;
+            this.executor = executor;
+            this.updateScheduledExecutor = updateScheduledExecutor;
+            List<BufferInfo> bufferStates = outputBuffers.getBuffers()
+                    .keySet().stream()
+                    .map(outputId -> new BufferInfo(outputId, false, 0, 0, PageBufferInfo.empty()))
+                    .collect(toImmutableList());
+            this.taskInfo = createInitialTask(taskId, location, bufferStates, new TaskStats(DateTime.now(), null), nodeId);
+            this.taskUpdateRequestCodec = taskUpdateRequestCodec;
+            this.planFragmentCodec = planFragmentCodec;
+            this.taskInfoCodec = taskInfoCodec;
+        }
+    }
+
+    public TaskInfo getTaskInfo()
+    {
+        return taskInfo;
+    }
+
+    public void start()
+    {
+        sendUpdate();
+        scheduleGetTaskInfo();
+    }
+
```

Comment:
I'm wondering the usage of this class. Are we supposed to let the user call start() to start the task and then continuously calling getTaskInfo() to get status of the task and when TaskInfo shows task finished then user is then responsible to call stop()?

If that is the case I wonder if we should give user an explicit signal to tell that task has finished, that is maybe return a future from start() call. But it's up to you and @miaoever

Reply:
Agree, from our offline discussion, the `getTaskInfo()` should return a `Future` to the caller.

Reply:
Capture the offline discussion here: we decide to add that in a separate PR